File: pkg/core/state_advancer.go
package core

import (
	"fmt"
)

var _ StateAdvancer = &StateAdvancerV1{}

// StateAdvancerV1 provides basic validation for state transitions
type StateAdvancerV1 struct {
	assetStore AssetStore
}

// NewStateAdvancerV1 creates a new simple transition validator
func NewStateAdvancerV1(assetStore AssetStore) *StateAdvancerV1 {
	return &StateAdvancerV1{
		assetStore: assetStore,
	}
}

// ValidateAdvancement validates that the proposed state is a valid advancement of the current state
//
// NOTE: User signature is not validated here
//
// TODO: Add shared JSON fixture suite consumed by both Go and TS test suites to guarantee validation parity
func (v *StateAdvancerV1) ValidateAdvancement(currentState, proposedState State) error {
	expectedState := currentState.NextState()
	if proposedState.HomeChannelID == nil {
		return fmt.Errorf("home channel ID cannot be nil")
	}

	// A non-nil state with a nil HomeChannelID is valid — it represents a user who received
	// funds but has not yet opened a channel. In that case, adopt the proposed state's channel info.
	if expectedState.HomeChannelID == nil {
		expectedState.HomeChannelID = proposedState.HomeChannelID
		expectedState.HomeLedger.BlockchainID = proposedState.HomeLedger.BlockchainID
		expectedState.HomeLedger.TokenAddress = proposedState.HomeLedger.TokenAddress
	}

	if *proposedState.HomeChannelID != *expectedState.HomeChannelID {
		return fmt.Errorf("home channel ID mismatch: expected=%s, proposed=%s", *expectedState.HomeChannelID, *proposedState.HomeChannelID)
	}

	// Version must increment
	if proposedState.Version != expectedState.Version {
		return fmt.Errorf("version mismatch: expected=%d, proposed=%d", expectedState.Version, proposedState.Version)
	}

	// User wallet must match
	if proposedState.UserWallet != expectedState.UserWallet {
		return fmt.Errorf("user wallet mismatch: expected=%s, proposed=%s", expectedState.UserWallet, proposedState.UserWallet)
	}

	// Asset must match
	if proposedState.Asset != expectedState.Asset {
		return fmt.Errorf("asset mismatch: expected=%s, proposed=%s", expectedState.Asset, proposedState.Asset)
	}

	// Epoch must match
	if proposedState.Epoch != expectedState.Epoch {
		return fmt.Errorf("epoch mismatch: expected=%d, proposed=%d", expectedState.Epoch, proposedState.Epoch)
	}

	// State ID must match
	if proposedState.ID != expectedState.ID {
		return fmt.Errorf("state ID mismatch: expected=%s, proposed=%s", expectedState.ID, proposedState.ID)
	}

	newTransition := proposedState.Transition

	decimals, err := v.assetStore.GetAssetDecimals(proposedState.Asset)
	if err != nil {
		return fmt.Errorf("failed to get asset decimals: %w", err)
	}

	if err := ValidateDecimalPrecision(newTransition.Amount, decimals); err != nil {
		return fmt.Errorf("invalid amount for asset %s: %w", proposedState.Asset, err)
	}

	switch newTransition.Type {
	case TransitionTypeAcknowledgement:
		if !newTransition.Amount.IsZero() {
			return fmt.Errorf("transition amount must be zero, got %s", newTransition.Amount.String())
		}
	case TransitionTypeFinalize:
		if newTransition.Amount.IsNegative() {
			return fmt.Errorf("transition amount must not be negative, got %s", newTransition.Amount.String())
		}
	default:
		if !newTransition.Amount.IsPositive() {
			return fmt.Errorf("transition amount must be positive, got %s", newTransition.Amount.String())
		}
	}

	lastTransition := currentState.Transition

	switch newTransition.Type {
	case TransitionTypeVoid:
		return fmt.Errorf("cannot apply void transition as new transition")
	case TransitionTypeAcknowledgement:
		if currentState.UserSig != nil {
			return fmt.Errorf("current state is already acknowledged")
		}
		_, err = expectedState.ApplyAcknowledgementTransition()
	case TransitionTypeHomeDeposit:
		_, err = expectedState.ApplyHomeDepositTransition(newTransition.Amount)
	case TransitionTypeHomeWithdrawal:
		_, err = expectedState.ApplyHomeWithdrawalTransition(newTransition.Amount)
	case TransitionTypeTransferSend:
		_, err = expectedState.ApplyTransferSendTransition(newTransition.AccountID, newTransition.Amount)
	case TransitionTypeCommit:
		_, err = expectedState.ApplyCommitTransition(newTransition.AccountID, newTransition.Amount)
	case TransitionTypeMutualLock:
		if proposedState.EscrowLedger == nil {
			return fmt.Errorf("proposed state escrow ledger is nil")
		}
		_, err = expectedState.ApplyMutualLockTransition(
			proposedState.EscrowLedger.BlockchainID,
			proposedState.EscrowLedger.TokenAddress,
			newTransition.Amount)
	case TransitionTypeEscrowDeposit:
		if lastTransition.Type == TransitionTypeMutualLock {
			if !lastTransition.Amount.Equal(newTransition.Amount) {
				return fmt.Errorf("escrow deposit amount must be the same as mutual lock amount")
			}
			_, err = expectedState.ApplyEscrowDepositTransition(newTransition.Amount)
		} else {
			return fmt.Errorf("escrow deposit transition must follow a mutual lock transition")
		}
	case TransitionTypeEscrowLock:
		if proposedState.EscrowLedger == nil {
			return fmt.Errorf("proposed state escrow ledger is nil")
		}
		_, err = expectedState.ApplyEscrowLockTransition(
			proposedState.EscrowLedger.BlockchainID,
			proposedState.EscrowLedger.TokenAddress,
			newTransition.Amount)
	case TransitionTypeEscrowWithdraw:
		if lastTransition.Type == TransitionTypeEscrowLock {
			if !lastTransition.Amount.Equal(newTransition.Amount) {
				return fmt.Errorf("escrow withdraw amount must be the same as escrow lock amount")
			}
			_, err = expectedState.ApplyEscrowWithdrawTransition(newTransition.Amount)
		} else {
			return fmt.Errorf("escrow withdraw transition must follow an escrow lock transition")
		}
	case TransitionTypeMigrate:
		_, err = expectedState.ApplyMigrateTransition(newTransition.Amount)
	case TransitionTypeFinalize:
		_, err = expectedState.ApplyFinalizeTransition()

	default:
		return fmt.Errorf("unsupported type for new transition: %d", newTransition.Type)
	}
	if err != nil {
		return fmt.Errorf("failed to apply new transition: %w", err)
	}

	expectedTransition := expectedState.Transition
	if err := expectedTransition.Equal(newTransition); err != nil {
		return fmt.Errorf("new transition does not match expected: %w", err)
	}

	if err := proposedState.HomeLedger.Equal(expectedState.HomeLedger); err != nil {
		return fmt.Errorf("home ledger mismatch: %w", err)
	}
	if err := proposedState.HomeLedger.Validate(); err != nil {
		return fmt.Errorf("invalid home ledger: %w", err)
	}

	if (expectedState.EscrowChannelID == nil) != (proposedState.EscrowChannelID == nil) {
		return fmt.Errorf("escrow channel ID presence mismatch")
	}

	if expectedState.EscrowChannelID != nil && proposedState.EscrowChannelID != nil {
		if *expectedState.EscrowChannelID != *proposedState.EscrowChannelID {
			return fmt.Errorf("escrow channel ID mismatch: expected=%s, proposed=%s", *expectedState.EscrowChannelID, *proposedState.EscrowChannelID)
		}
	}

	if (expectedState.EscrowLedger == nil) != (proposedState.EscrowLedger == nil) {
		return fmt.Errorf("escrow ledger presence mismatch")
	}

	if expectedState.EscrowLedger != nil && proposedState.EscrowLedger != nil {
		if err := proposedState.EscrowLedger.Equal(*expectedState.EscrowLedger); err != nil {
			return fmt.Errorf("escrow ledger mismatch: %w", err)
		}
		if err := proposedState.EscrowLedger.Validate(); err != nil {
			return fmt.Errorf("invalid escrow ledger: %w", err)
		}

		if proposedState.EscrowLedger.BlockchainID == proposedState.HomeLedger.BlockchainID {
			return fmt.Errorf("escrow ledger blockchain ID cannot match home ledger blockchain ID")
		}
	}

	return nil
}

File: sdk/ts/src/core/state_advancer.ts
import { State, TransitionType, ledgerEqual, validateLedger, transitionsEqual } from './types.js';
import { StateAdvancer, AssetStore } from './interface.js';
import { validateDecimalPrecision } from './utils.js';
import {
  nextState,
  applyAcknowledgementTransition,
  applyHomeDepositTransition,
  applyHomeWithdrawalTransition,
  applyTransferSendTransition,
  applyCommitTransition,
  applyMutualLockTransition,
  applyEscrowDepositTransition,
  applyEscrowLockTransition,
  applyEscrowWithdrawTransition,
  applyMigrateTransition,
  applyFinalizeTransition,
} from './state.js';

export class StateAdvancerV1 implements StateAdvancer {
  constructor(private assetStore: AssetStore) {}

  // ValidateAdvancement validates that the proposed state is a valid advancement of the current state
  //
  // NOTE: User signature is not validated here
  //
  // TODO: Add shared JSON fixture suite consumed by both Go and TS test suites to guarantee validation parity
  async validateAdvancement(currentState: State, proposedState: State): Promise<void> {
    const expectedState = nextState(currentState);

    if (!proposedState.homeChannelId) {
      throw new Error('home channel ID cannot be nil');
    }

    if (!expectedState.homeChannelId) {
      expectedState.homeChannelId = proposedState.homeChannelId;
      expectedState.homeLedger.blockchainId = proposedState.homeLedger.blockchainId;
      expectedState.homeLedger.tokenAddress = proposedState.homeLedger.tokenAddress;
    }

    if (proposedState.homeChannelId !== expectedState.homeChannelId) {
      throw new Error(
        `home channel ID mismatch: expected=${expectedState.homeChannelId}, proposed=${proposedState.homeChannelId}`
      );
    }

    if (proposedState.version !== expectedState.version) {
      throw new Error(
        `version mismatch: expected=${expectedState.version}, proposed=${proposedState.version}`
      );
    }

    if (proposedState.userWallet.toLowerCase() !== expectedState.userWallet.toLowerCase()) {
      throw new Error(
        `user wallet mismatch: expected=${expectedState.userWallet}, proposed=${proposedState.userWallet}`
      );
    }

    if (proposedState.asset !== expectedState.asset) {
      throw new Error(
        `asset mismatch: expected=${expectedState.asset}, proposed=${proposedState.asset}`
      );
    }

    if (proposedState.epoch !== expectedState.epoch) {
      throw new Error(
        `epoch mismatch: expected=${expectedState.epoch}, proposed=${proposedState.epoch}`
      );
    }

    if (proposedState.id !== expectedState.id) {
      throw new Error(
        `state ID mismatch: expected=${expectedState.id}, proposed=${proposedState.id}`
      );
    }

    const newTransition = proposedState.transition;

    const decimals = await this.assetStore.getAssetDecimals(proposedState.asset);
    validateDecimalPrecision(newTransition.amount, decimals);

    switch (newTransition.type) {
      case TransitionType.Acknowledgement:
        if (!newTransition.amount.isZero()) {
          throw new Error(
            `transition amount must be zero, got ${newTransition.amount.toString()}`
          );
        }
        break;
      case TransitionType.Finalize:
        if (newTransition.amount.isNegative()) {
          throw new Error(
            `transition amount must not be negative, got ${newTransition.amount.toString()}`
          );
        }
        break;
      default:
        if (newTransition.amount.isNegative() || newTransition.amount.isZero()) {
          throw new Error(
            `transition amount must be positive, got ${newTransition.amount.toString()}`
          );
        }
    }

    const lastTransition = currentState.transition;

    switch (newTransition.type) {
      case TransitionType.Void:
        throw new Error('cannot apply void transition as new transition');
      case TransitionType.Acknowledgement:
        if (currentState.userSig) {
          throw new Error('current state is already acknowledged');
        }
        applyAcknowledgementTransition(expectedState);
        break;
      case TransitionType.HomeDeposit:
        applyHomeDepositTransition(expectedState, newTransition.amount);
        break;
      case TransitionType.HomeWithdrawal:
        applyHomeWithdrawalTransition(expectedState, newTransition.amount);
        break;
      case TransitionType.TransferSend:
        if (!newTransition.accountId) {
          throw new Error('missing account ID for transfer send transition');
        }
        applyTransferSendTransition(expectedState, newTransition.accountId, newTransition.amount);
        break;
      case TransitionType.Commit:
        if (!newTransition.accountId) {
          throw new Error('missing account ID for commit transition');
        }
        applyCommitTransition(expectedState, newTransition.accountId, newTransition.amount);
        break;
      case TransitionType.MutualLock:
        if (!proposedState.escrowLedger) {
          throw new Error('proposed state escrow ledger is nil');
        }
        applyMutualLockTransition(
          expectedState,
          proposedState.escrowLedger.blockchainId,
          proposedState.escrowLedger.tokenAddress,
          newTransition.amount
        );
        break;
      case TransitionType.EscrowDeposit:
        if (lastTransition.type === TransitionType.MutualLock) {
          if (!lastTransition.amount.equals(newTransition.amount)) {
            throw new Error('escrow deposit amount must be the same as mutual lock amount');
          }
          applyEscrowDepositTransition(expectedState, newTransition.amount);
        } else {
          throw new Error('escrow deposit transition must follow a mutual lock transition');
        }
        break;
      case TransitionType.EscrowLock:
        if (!proposedState.escrowLedger) {
          throw new Error('proposed state escrow ledger is nil');
        }
        applyEscrowLockTransition(
          expectedState,
          proposedState.escrowLedger.blockchainId,
          proposedState.escrowLedger.tokenAddress,
          newTransition.amount
        );
        break;
      case TransitionType.EscrowWithdraw:
        if (lastTransition.type === TransitionType.EscrowLock) {
          if (!lastTransition.amount.equals(newTransition.amount)) {
            throw new Error('escrow withdraw amount must be the same as escrow lock amount');
          }
          applyEscrowWithdrawTransition(expectedState, newTransition.amount);
        } else {
          throw new Error('escrow withdraw transition must follow an escrow lock transition');
        }
        break;
      case TransitionType.Migrate:
        applyMigrateTransition(expectedState, newTransition.amount);
        break;
      case TransitionType.Finalize:
        applyFinalizeTransition(expectedState);
        break;
      default:
        throw new Error(`unsupported type for new transition: ${newTransition.type}`);
    }

    const expectedTransition = expectedState.transition;
    const transitionMismatch = transitionsEqual(expectedTransition, newTransition);
    if (transitionMismatch) {
      throw new Error(`new transition does not match expected: ${transitionMismatch}`);
    }

    const homeLedgerMismatch = ledgerEqual(expectedState.homeLedger, proposedState.homeLedger);
    if (homeLedgerMismatch) {
      throw new Error(`home ledger mismatch: ${homeLedgerMismatch}`);
    }
    validateLedger(proposedState.homeLedger);

    const expectedHasEscrowId = expectedState.escrowChannelId !== undefined;
    const proposedHasEscrowId = proposedState.escrowChannelId !== undefined;
    if (expectedHasEscrowId !== proposedHasEscrowId) {
      throw new Error('escrow channel ID presence mismatch');
    }

    if (expectedState.escrowChannelId && proposedState.escrowChannelId) {
      if (expectedState.escrowChannelId !== proposedState.escrowChannelId) {
        throw new Error(
          `escrow channel ID mismatch: expected=${expectedState.escrowChannelId}, proposed=${proposedState.escrowChannelId}`
        );
      }
    }

    const expectedHasEscrowLedger = expectedState.escrowLedger !== undefined;
    const proposedHasEscrowLedger = proposedState.escrowLedger !== undefined;
    if (expectedHasEscrowLedger !== proposedHasEscrowLedger) {
      throw new Error('escrow ledger presence mismatch');
    }

    if (expectedState.escrowLedger && proposedState.escrowLedger) {
      const escrowLedgerMismatch = ledgerEqual(
        expectedState.escrowLedger,
        proposedState.escrowLedger
      );
      if (escrowLedgerMismatch) {
        throw new Error(`escrow ledger mismatch: expected=${JSON.stringify(expectedState.escrowLedger)}, proposed=${JSON.stringify(proposedState.escrowLedger)}: ${escrowLedgerMismatch}`);
      }
      validateLedger(proposedState.escrowLedger);

      if (proposedState.escrowLedger.blockchainId === proposedState.homeLedger.blockchainId) {
        throw new Error('escrow ledger blockchain ID cannot match home ledger blockchain ID');
      }
    }
  }
}
